core/src/main/java/org/apache/iceberg/actions/SortStrategy.java

```diff
@@ -0,0 +1,140 @@
+ * in relation to a column. For example, if the Sort strategy is used on a set of files which is ordered
+ * by column x and original has files File A (x: 0 - 50), File B ( x: 10 - 40) and File C ( x: 30 - 60),
+ * this Strategy will attempt to rewrite those files into File A' (x: 0-20), File B' (x: 21 - 40),
+ * File C' (x: 41 - 60).
+ * <p>
+ * Currently the there is no clustering detection and we will rewrite all files if {@link SortStrategy#REWRITE_ALL}
+ * is true (default: false). If this property is disabled any files with the incorrect sort-order as well as any files
+ * that would be chosen by {@link BinPackStrategy} will be rewrite candidates.
+ * <p>
+ * In the future other algorithms for determining files to rewrite will be provided.
+ */
+public abstract class SortStrategy extends BinPackStrategy {
+  private static final Logger LOG = LoggerFactory.getLogger(SortStrategy.class);
+
+  /**
+   * Rewrites all files, regardless of their size. Defaults to false, rewriting only wrong sort-order and mis-sized
+   * files;
+   */
+  public static final String REWRITE_ALL = "no-size-filter";
+  public static final boolean REWRITE_ALL_DEFAULT = false;
+
+  private static final Set<String> validOptions = ImmutableSet.of(
+      REWRITE_ALL
+  );
+
+  private boolean rewriteAll;
+  private SortOrder sortOrder;
+  private int sortOrderId = -1;
+
+  /**
+   * Sets the sort order to be used in this strategy when rewriting files
+   * @param order the order to use
+   * @return this for method chaining
+   */
+  public SortStrategy sortOrder(SortOrder order) {
+    this.sortOrder = order;
+
+    // See if this order matches any of our known orders
+    Optional<Entry<Integer, SortOrder>> knownOrder = table().sortOrders().entrySet().stream()
+        .filter(entry -> entry.getValue().sameOrder(order))
+        .findFirst();
+    knownOrder.ifPresent(entry -> sortOrderId = entry.getKey());
+
+    return this;
+  }
+
+  @Override
+  public String name() {
+    return "SORT";
+  }
+
+  @Override
+  public Set<String> validOptions() {
+    return ImmutableSet.<String>builder()
+        .addAll(super.validOptions())
+        .addAll(validOptions)
+        .build();
+  }
+
+  @Override
+  public RewriteStrategy options(Map<String, String> options) {
+    super.options(options); // Also checks validity of BinPack options
+
+    rewriteAll = PropertyUtil.propertyAsBoolean(options,
+        REWRITE_ALL,
+        REWRITE_ALL_DEFAULT);
+
+    if (sortOrder == null) {
+      sortOrder = table().sortOrder();
+      sortOrderId = sortOrder.orderId();
+    }
+
+    validateOptions();
+    return this;
+  }
+
+  @Override
+  public Iterable<FileScanTask> selectFilesToRewrite(Iterable<FileScanTask> dataFiles) {
+    if (rewriteAll) {
+      LOG.info("Sort Strategy for table {} set to rewrite all data files", table().name());
+      return dataFiles;
+    } else {
+      FluentIterable filesWithCorrectOrder =
```

Comment:
I think the current implementation as it stands today will always rewrite everything as we never currently propagate the sort order id to data files.